{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7dbf5833-ff57-43c7-bb4a-897cf1cf5bea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              sseqid\n",
      "0         OV277454.1\n",
      "1         MN894805.1\n",
      "2         CP076413.1\n",
      "3     XR_006695749.1\n",
      "4         OV277454.1\n",
      "...              ...\n",
      "9331     NR_103099.1\n",
      "9332  XR_006269445.1\n",
      "9333      KF104644.1\n",
      "9334      CP068261.2\n",
      "9335      AL590668.5\n",
      "\n",
      "[9336 rows x 1 columns]\n",
      "Empty DataFrame\n",
      "Columns: [sseqid]\n",
      "Index: []\n",
      "OV277454.1        5223\n",
      "MN894805.1         586\n",
      "XR_006695749.1     152\n",
      "OV277463.1         151\n",
      "OV277457.1         108\n",
      "                  ... \n",
      "OV179167.1           1\n",
      "HQ659117.1           1\n",
      "AC254813.1           1\n",
      "CP087586.1           1\n",
      "AL590668.5           1\n",
      "Name: sseqid, Length: 1290, dtype: int64\n",
      "Series([], Name: sseqid, dtype: int64)\n",
      "9336\n",
      "0\n",
      "XR_006695749.1\n",
      "OV277463.1\n",
      "OV277457.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_34618/640869443.py:256: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  sseq_count = sseq_count_series3.append(ser)\n",
      "/tmp/ipykernel_34618/640869443.py:291: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  sseq_count2 = sseq_count_series6.append(ser2)\n",
      "/tmp/ipykernel_34618/640869443.py:324: RuntimeWarning: invalid value encountered in long_scalars\n",
      "  wedges2, *_ = ax3.pie(count2, labels=[f'{sseq2[i]}: {count2[i]} ({count2[i]/sum(count2)*100:.1f}% )' for i in range(len(sseq2))], colors=None,autopct='',startangle=45,\n",
      "/opt/conda/lib/python3.9/site-packages/matplotlib/axes/_axes.py:3050: RuntimeWarning: invalid value encountered in true_divide\n",
      "  x = x / sx\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                              sseqid  count\n",
      "0        Meles meles genome assembly, chromosome: 13   5223\n",
      "1  Homo sapiens haplogroup H3v+16093 mitochondrio...    586\n",
      "2                                              Other   3527\n",
      "  sseqid  count\n",
      "0  Other      0\n",
      "[152, 151, 108]\n",
      "[]\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "need at least one array to concatenate",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [19]\u001b[0m, in \u001b[0;36m<cell line: 364>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    359\u001b[0m ax5\u001b[38;5;241m.\u001b[39mtext(\u001b[38;5;241m0.95\u001b[39m, \u001b[38;5;241m2.2\u001b[39m, text1, transform\u001b[38;5;241m=\u001b[39max6\u001b[38;5;241m.\u001b[39mtransAxes, fontsize\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m14\u001b[39m,\n\u001b[1;32m    360\u001b[0m         verticalalignment\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtop\u001b[39m\u001b[38;5;124m'\u001b[39m, bbox\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mdict\u001b[39m(facecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m, edgecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mblack\u001b[39m\u001b[38;5;124m'\u001b[39m, boxstyle\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mround,pad=1\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m    361\u001b[0m ax6\u001b[38;5;241m.\u001b[39mtext(\u001b[38;5;241m0.95\u001b[39m, \u001b[38;5;241m0.95\u001b[39m, text2, transform\u001b[38;5;241m=\u001b[39max6\u001b[38;5;241m.\u001b[39mtransAxes, fontsize\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m14\u001b[39m,\n\u001b[1;32m    362\u001b[0m         verticalalignment\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtop\u001b[39m\u001b[38;5;124m'\u001b[39m, bbox\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mdict\u001b[39m(facecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m, edgecolor\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mblack\u001b[39m\u001b[38;5;124m'\u001b[39m, boxstyle\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mround,pad=1\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[0;32m--> 364\u001b[0m \u001b[43mfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtight_layout\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5.0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    365\u001b[0m plt\u001b[38;5;241m.\u001b[39mshow(block\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/figure.py:3224\u001b[0m, in \u001b[0;36mFigure.tight_layout\u001b[0;34m(self, pad, h_pad, w_pad, rect)\u001b[0m\n\u001b[1;32m   3222\u001b[0m renderer \u001b[38;5;241m=\u001b[39m get_renderer(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m   3223\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(renderer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_draw_disabled\u001b[39m\u001b[38;5;124m\"\u001b[39m, nullcontext)():\n\u001b[0;32m-> 3224\u001b[0m     kwargs \u001b[38;5;241m=\u001b[39m \u001b[43mget_tight_layout_figure\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3225\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maxes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msubplotspec_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3226\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpad\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mh_pad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mh_pad\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mw_pad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mw_pad\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrect\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3227\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwargs:\n\u001b[1;32m   3228\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msubplots_adjust(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/tight_layout.py:320\u001b[0m, in \u001b[0;36mget_tight_layout_figure\u001b[0;34m(fig, axes_list, subplotspec_list, renderer, pad, h_pad, w_pad, rect)\u001b[0m\n\u001b[1;32m    315\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m {}\n\u001b[1;32m    316\u001b[0m     span_pairs\u001b[38;5;241m.\u001b[39mappend((\n\u001b[1;32m    317\u001b[0m         \u001b[38;5;28mslice\u001b[39m(ss\u001b[38;5;241m.\u001b[39mrowspan\u001b[38;5;241m.\u001b[39mstart \u001b[38;5;241m*\u001b[39m div_row, ss\u001b[38;5;241m.\u001b[39mrowspan\u001b[38;5;241m.\u001b[39mstop \u001b[38;5;241m*\u001b[39m div_row),\n\u001b[1;32m    318\u001b[0m         \u001b[38;5;28mslice\u001b[39m(ss\u001b[38;5;241m.\u001b[39mcolspan\u001b[38;5;241m.\u001b[39mstart \u001b[38;5;241m*\u001b[39m div_col, ss\u001b[38;5;241m.\u001b[39mcolspan\u001b[38;5;241m.\u001b[39mstop \u001b[38;5;241m*\u001b[39m div_col)))\n\u001b[0;32m--> 320\u001b[0m kwargs \u001b[38;5;241m=\u001b[39m \u001b[43m_auto_adjust_subplotpars\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    321\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mshape\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mmax_nrows\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_ncols\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    322\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mspan_pairs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mspan_pairs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    323\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43msubplot_list\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msubplot_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    324\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43max_bbox_list\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43max_bbox_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    325\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mpad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpad\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mh_pad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mh_pad\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mw_pad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mw_pad\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    327\u001b[0m \u001b[38;5;66;03m# kwargs can be none if tight_layout fails...\u001b[39;00m\n\u001b[1;32m    328\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m rect \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m kwargs \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    329\u001b[0m     \u001b[38;5;66;03m# if rect is given, the whole subplots area (including\u001b[39;00m\n\u001b[1;32m    330\u001b[0m     \u001b[38;5;66;03m# labels) will fit into the rect instead of the\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    334\u001b[0m     \u001b[38;5;66;03m# auto_adjust_subplotpars twice, where the second run\u001b[39;00m\n\u001b[1;32m    335\u001b[0m     \u001b[38;5;66;03m# with adjusted rect parameters.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/tight_layout.py:82\u001b[0m, in \u001b[0;36m_auto_adjust_subplotpars\u001b[0;34m(fig, renderer, shape, span_pairs, subplot_list, ax_bbox_list, pad, h_pad, w_pad, rect)\u001b[0m\n\u001b[1;32m     80\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ax\u001b[38;5;241m.\u001b[39mget_visible():\n\u001b[1;32m     81\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 82\u001b[0m         bb \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m [\u001b[43max\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_tightbbox\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfor_layout_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m]\n\u001b[1;32m     83\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m     84\u001b[0m         bb \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m [ax\u001b[38;5;241m.\u001b[39mget_tightbbox(renderer)]\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/axes/_base.py:4675\u001b[0m, in \u001b[0;36m_AxesBase.get_tightbbox\u001b[0;34m(self, renderer, call_axes_locator, bbox_extra_artists, for_layout_only)\u001b[0m\n\u001b[1;32m   4671\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m np\u001b[38;5;241m.\u001b[39mall(clip_extent\u001b[38;5;241m.\u001b[39mextents \u001b[38;5;241m==\u001b[39m axbbox\u001b[38;5;241m.\u001b[39mextents):\n\u001b[1;32m   4672\u001b[0m         \u001b[38;5;66;03m# clip extent is inside the Axes bbox so don't check\u001b[39;00m\n\u001b[1;32m   4673\u001b[0m         \u001b[38;5;66;03m# this artist\u001b[39;00m\n\u001b[1;32m   4674\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[0;32m-> 4675\u001b[0m bbox \u001b[38;5;241m=\u001b[39m \u001b[43ma\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_tightbbox\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4676\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (bbox \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   4677\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;241m<\u001b[39m bbox\u001b[38;5;241m.\u001b[39mwidth \u001b[38;5;241m<\u001b[39m np\u001b[38;5;241m.\u001b[39minf\n\u001b[1;32m   4678\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;241m<\u001b[39m bbox\u001b[38;5;241m.\u001b[39mheight \u001b[38;5;241m<\u001b[39m np\u001b[38;5;241m.\u001b[39minf):\n\u001b[1;32m   4679\u001b[0m     bb\u001b[38;5;241m.\u001b[39mappend(bbox)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/artist.py:355\u001b[0m, in \u001b[0;36mArtist.get_tightbbox\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m    340\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_tightbbox\u001b[39m(\u001b[38;5;28mself\u001b[39m, renderer):\n\u001b[1;32m    341\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    342\u001b[0m \u001b[38;5;124;03m    Like `.Artist.get_window_extent`, but includes any clipping.\u001b[39;00m\n\u001b[1;32m    343\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    353\u001b[0m \u001b[38;5;124;03m        The enclosing bounding box (in figure pixel coordinates).\u001b[39;00m\n\u001b[1;32m    354\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 355\u001b[0m     bbox \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_window_extent\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    356\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_clip_on():\n\u001b[1;32m    357\u001b[0m         clip_box \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_clip_box()\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/patches.py:629\u001b[0m, in \u001b[0;36mPatch.get_window_extent\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m    628\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_window_extent\u001b[39m(\u001b[38;5;28mself\u001b[39m, renderer\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 629\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_path\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_extents\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/path.py:631\u001b[0m, in \u001b[0;36mPath.get_extents\u001b[0;34m(self, transform, **kwargs)\u001b[0m\n\u001b[1;32m    629\u001b[0m         \u001b[38;5;66;03m# as can the ends of the curve\u001b[39;00m\n\u001b[1;32m    630\u001b[0m         xys\u001b[38;5;241m.\u001b[39mappend(curve([\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m*\u001b[39mdzeros, \u001b[38;5;241m1\u001b[39m]))\n\u001b[0;32m--> 631\u001b[0m     xys \u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconcatenate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mxys\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    632\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(xys):\n\u001b[1;32m    633\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m Bbox([xys\u001b[38;5;241m.\u001b[39mmin(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m), xys\u001b[38;5;241m.\u001b[39mmax(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)])\n",
      "File \u001b[0;32m<__array_function__ internals>:5\u001b[0m, in \u001b[0;36mconcatenate\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: need at least one array to concatenate"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "posx and posy should be finite values\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "need at least one array to concatenate",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/IPython/core/formatters.py:339\u001b[0m, in \u001b[0;36mBaseFormatter.__call__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    337\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m    338\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 339\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mprinter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    340\u001b[0m \u001b[38;5;66;03m# Finally look for special method names\u001b[39;00m\n\u001b[1;32m    341\u001b[0m method \u001b[38;5;241m=\u001b[39m get_real_method(obj, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprint_method)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/IPython/core/pylabtools.py:151\u001b[0m, in \u001b[0;36mprint_figure\u001b[0;34m(fig, fmt, bbox_inches, base64, **kwargs)\u001b[0m\n\u001b[1;32m    148\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbackend_bases\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m FigureCanvasBase\n\u001b[1;32m    149\u001b[0m     FigureCanvasBase(fig)\n\u001b[0;32m--> 151\u001b[0m \u001b[43mfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcanvas\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprint_figure\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbytes_io\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    152\u001b[0m data \u001b[38;5;241m=\u001b[39m bytes_io\u001b[38;5;241m.\u001b[39mgetvalue()\n\u001b[1;32m    153\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m fmt \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msvg\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/backend_bases.py:2299\u001b[0m, in \u001b[0;36mFigureCanvasBase.print_figure\u001b[0;34m(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)\u001b[0m\n\u001b[1;32m   2297\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m bbox_inches:\n\u001b[1;32m   2298\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m bbox_inches \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtight\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 2299\u001b[0m         bbox_inches \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_tightbbox\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2300\u001b[0m \u001b[43m            \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbbox_extra_artists\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbbox_extra_artists\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2301\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m pad_inches \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   2302\u001b[0m             pad_inches \u001b[38;5;241m=\u001b[39m rcParams[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msavefig.pad_inches\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/figure.py:1684\u001b[0m, in \u001b[0;36mFigureBase.get_tightbbox\u001b[0;34m(self, renderer, bbox_extra_artists)\u001b[0m\n\u001b[1;32m   1681\u001b[0m     artists \u001b[38;5;241m=\u001b[39m bbox_extra_artists\n\u001b[1;32m   1683\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m a \u001b[38;5;129;01min\u001b[39;00m artists:\n\u001b[0;32m-> 1684\u001b[0m     bbox \u001b[38;5;241m=\u001b[39m \u001b[43ma\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_tightbbox\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1685\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m bbox \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m (bbox\u001b[38;5;241m.\u001b[39mwidth \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m bbox\u001b[38;5;241m.\u001b[39mheight \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m):\n\u001b[1;32m   1686\u001b[0m         bb\u001b[38;5;241m.\u001b[39mappend(bbox)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/axes/_base.py:4675\u001b[0m, in \u001b[0;36m_AxesBase.get_tightbbox\u001b[0;34m(self, renderer, call_axes_locator, bbox_extra_artists, for_layout_only)\u001b[0m\n\u001b[1;32m   4671\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m np\u001b[38;5;241m.\u001b[39mall(clip_extent\u001b[38;5;241m.\u001b[39mextents \u001b[38;5;241m==\u001b[39m axbbox\u001b[38;5;241m.\u001b[39mextents):\n\u001b[1;32m   4672\u001b[0m         \u001b[38;5;66;03m# clip extent is inside the Axes bbox so don't check\u001b[39;00m\n\u001b[1;32m   4673\u001b[0m         \u001b[38;5;66;03m# this artist\u001b[39;00m\n\u001b[1;32m   4674\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[0;32m-> 4675\u001b[0m bbox \u001b[38;5;241m=\u001b[39m \u001b[43ma\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_tightbbox\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4676\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (bbox \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   4677\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;241m<\u001b[39m bbox\u001b[38;5;241m.\u001b[39mwidth \u001b[38;5;241m<\u001b[39m np\u001b[38;5;241m.\u001b[39minf\n\u001b[1;32m   4678\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;241m<\u001b[39m bbox\u001b[38;5;241m.\u001b[39mheight \u001b[38;5;241m<\u001b[39m np\u001b[38;5;241m.\u001b[39minf):\n\u001b[1;32m   4679\u001b[0m     bb\u001b[38;5;241m.\u001b[39mappend(bbox)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/artist.py:355\u001b[0m, in \u001b[0;36mArtist.get_tightbbox\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m    340\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_tightbbox\u001b[39m(\u001b[38;5;28mself\u001b[39m, renderer):\n\u001b[1;32m    341\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    342\u001b[0m \u001b[38;5;124;03m    Like `.Artist.get_window_extent`, but includes any clipping.\u001b[39;00m\n\u001b[1;32m    343\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    353\u001b[0m \u001b[38;5;124;03m        The enclosing bounding box (in figure pixel coordinates).\u001b[39;00m\n\u001b[1;32m    354\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 355\u001b[0m     bbox \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_window_extent\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    356\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_clip_on():\n\u001b[1;32m    357\u001b[0m         clip_box \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_clip_box()\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/patches.py:629\u001b[0m, in \u001b[0;36mPatch.get_window_extent\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m    628\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_window_extent\u001b[39m(\u001b[38;5;28mself\u001b[39m, renderer\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 629\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_path\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_extents\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/matplotlib/path.py:631\u001b[0m, in \u001b[0;36mPath.get_extents\u001b[0;34m(self, transform, **kwargs)\u001b[0m\n\u001b[1;32m    629\u001b[0m         \u001b[38;5;66;03m# as can the ends of the curve\u001b[39;00m\n\u001b[1;32m    630\u001b[0m         xys\u001b[38;5;241m.\u001b[39mappend(curve([\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m*\u001b[39mdzeros, \u001b[38;5;241m1\u001b[39m]))\n\u001b[0;32m--> 631\u001b[0m     xys \u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconcatenate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mxys\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    632\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(xys):\n\u001b[1;32m    633\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m Bbox([xys\u001b[38;5;241m.\u001b[39mmin(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m), xys\u001b[38;5;241m.\u001b[39mmax(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)])\n",
      "File \u001b[0;32m<__array_function__ internals>:5\u001b[0m, in \u001b[0;36mconcatenate\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: need at least one array to concatenate"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 648x360 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#select intepreter and change environment to anaconda\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import urllib\n",
    "import xmltodict\n",
    "from xml.dom import minidom\n",
    "import time\n",
    "\n",
    "\n",
    "#handles command line input parameter\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# testing file is from 10,000 reads, so mapping percentage is expected to be out of 10,000\n",
    "# blast_tsv_file = sys.argv[1]\n",
    "# blast_tsv_file2 = sys.argv[2]\n",
    "# piePath = sys.argv[3]\n",
    "blast_tsv_file = 'LARP6.CTRL_IN1.umi.r1.fqdown_unmappedblast_downsampled_blastn.tsv'\n",
    "blast_tsv_file2 = 'LARP6.CTRL_IN1.umi.r1.fqdown_unmappedblast_downsampled_blastn.tsv'\n",
    "piePath = 'combined.png'\n",
    "piePathSample = os.path.splitext(piePath)[0]\n",
    "piePathSample2 = os.path.split(piePathSample)\n",
    "pieName = piePathSample2[1]\n",
    "\n",
    "\n",
    "#ncbi query functions\n",
    "def esearch(term, db='gds'):\n",
    "    \"\"\"\n",
    "    Queries NCBI using the esearch utility. GEO ('gds') database is used as default for search term.\n",
    "    \"\"\"\n",
    "    url = f'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db={db}&term={term}&retmax=5000&usehistory=y'\n",
    "    response = urllib.request.urlopen(url)\n",
    "    return response.read()\n",
    "\n",
    "\n",
    "def get_esummary(esearch_string, db='gds'):\n",
    "    \"\"\"\n",
    "    Parses a http response in XML format to obtain the webenv and querykey tokens.\n",
    "    Uses NCBI eutils to transform these tokens into web summaries of GEO (db='gds') datasets.\n",
    "    \"\"\"\n",
    "    xmldoc = minidom.parseString(esearch_string)\n",
    "    try:\n",
    "        webenv = xmldoc.getElementsByTagName('WebEnv')[0].firstChild.data\n",
    "        querykey = xmldoc.getElementsByTagName('QueryKey')[0].firstChild.data\n",
    "        host = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi'\n",
    "        params = f'?db={db}&version=2.0&query_key={querykey}&WebEnv={webenv}'\n",
    "        url = host + params\n",
    "        response = urllib.request.urlopen(url)\n",
    "        return response.read()\n",
    "    except IndexError as e:\n",
    "        print(f\"Unparsable publication string ({e}, search={esearch_string}\")\n",
    "        return \"\"\n",
    "\n",
    "#matplotlib to build the piechart\n",
    "fig, ((ax1, ax2, ax5), (ax3, ax4, ax6)) = plt.subplots(2, 3, figsize=(9, 5))\n",
    "fig.tight_layout(h_pad=2)\n",
    "plt.subplots_adjust(bottom=0, right=1.5, top=1.5)\n",
    "ax1.set_title('blastn piechart')\n",
    "ax2.set_title('blastn other')\n",
    "ax3.set_title('blastx piechart')\n",
    "ax4.set_title('blastx other')\n",
    "ax5.set_title('blastn percentage')\n",
    "ax6.set_title('blastx percentage')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Read the blast n file\n",
    "blast_N_tsv_Input = pd.read_csv(blast_tsv_file, header=None, sep='\\t')\n",
    "blast_N_tsv_Input.columns = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "\n",
    "df2 = blast_N_tsv_Input[['qseqid','sseqid','evalue']].copy()\n",
    "\n",
    "# Read the blast x file\n",
    "blast_X_tsv_Input = pd.read_csv(blast_tsv_file2, header=None, sep='\\t')\n",
    "blast_X_tsv_Input.columns = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### still need to reorganize the code for each different file below.\n",
    "\n",
    "\n",
    "#blast n output default is by best hit/lowest e value score, therefore add the first sseqid result for each qseqid\n",
    "# Get the unique sseqid and the corresponding qseqid, so we know which sequence is mapped by blasted.\n",
    "\n",
    "blastn_searchUnique_sseqid = []\n",
    "blastn_qseqid_found = []\n",
    "for idx in blast_N_tsv_Input.index:\n",
    "        if blast_N_tsv_Input['qseqid'][idx] not in blastn_qseqid_found:\n",
    "                blastn_qseqid_found.append(blast_N_tsv_Input['qseqid'][idx])\n",
    "                blastn_searchUnique_sseqid.append(blast_N_tsv_Input['sseqid'][idx])\n",
    "blastn_unique_sseqid = np.array(blastn_searchUnique_sseqid)\n",
    "\n",
    "#blast x output default is by best hit/lowest e value score, therefore add the first sseqid result for each qseqid\n",
    "# Get the unique sseqid and the corresponding qseqid, so we know which sequence is mapped by blasted.\n",
    "blastx_searchUnique_sseqid = []\n",
    "blastx_qseqid_found = []\n",
    "for idx in blast_X_tsv_Input.index:\n",
    "        if blast_X_tsv_Input['qseqid'][idx] not in blastx_qseqid_found:\n",
    "                blastx_qseqid_found.append(blast_X_tsv_Input['qseqid'][idx])\n",
    "blastx_unique_sseqid = np.array(blastx_searchUnique_sseqid)\n",
    "\n",
    "#Make a dataframe blast_n_sseqid_df for blast n sseqid and record the size of the mapped in blast_n_mapped_size\n",
    "blast_n_sseqid_df = pd.DataFrame(data = blastn_unique_sseqid, columns=['sseqid'])\n",
    "print(blast_n_sseqid_df)\n",
    "blast_n_mapped_size = blast_n_sseqid_df.size\n",
    "\n",
    "#Make a dataframe blast_x_sseqid_df for blast x sseqid and record the size of the mapped in blast_x_mapped_size\n",
    "blast_x_sseqid_df = pd.DataFrame(data = blastx_unique_sseqid, columns=['sseqid'])\n",
    "print(blast_x_sseqid_df)\n",
    "blast_x_mapped_size = blast_x_sseqid_df.size\n",
    "\n",
    "# Blast n\n",
    "sseq_count_series1 = blast_n_sseqid_df['sseqid'].value_counts()\n",
    "print(sseq_count_series1)\n",
    "\n",
    "# Blast x\n",
    "sseq_count_series4 = blast_x_sseqid_df['sseqid'].value_counts()\n",
    "print(sseq_count_series4)\n",
    "\n",
    "#loop through series to determine which elements to remove and add into \"other\" column\n",
    "# Blast n\n",
    "blast_n_genes_notTop_hit = []\n",
    "blast_n_otherCategory_size = 0\n",
    "blast_n_other_store = {}\n",
    "# Blast x\n",
    "blast_x_genes_notTop_hit = []\n",
    "blast_x_otherCategory_size = 0\n",
    "blast_x_other_store = {}\n",
    "\n",
    "#blast n number mapped\n",
    "print(blast_n_mapped_size)\n",
    "\n",
    "#blast x number mapped\n",
    "print(blast_x_mapped_size)\n",
    "\n",
    "# filtering out the sequence's top three mapped genes counts for blast n\n",
    "only_2_n = 2\n",
    "for index,values in sseq_count_series1.iteritems():\n",
    "        if(values < (0.001)*(blast_n_mapped_size)) or only_2_n == 0:\n",
    "                blast_n_genes_notTop_hit.append(index)\n",
    "                blast_n_otherCategory_size += values\n",
    "                blast_n_other_store[values] = index\n",
    "        else:\n",
    "            only_2_n -= 1\n",
    "\n",
    "# filtering out the sequence's top three mapped genes counts for blast x\n",
    "only_2_x = 2\n",
    "for index,values in sseq_count_series4.iteritems():\n",
    "        if(values < (0.001)*(blast_x_mapped_size)) or only_2_x == 0:\n",
    "                blast_x_genes_notTop_hit.append(index)\n",
    "                blast_x_otherCategory_size += values\n",
    "                blast_x_other_store[values] = index\n",
    "        else:\n",
    "            only_2_x -= 1\n",
    "\n",
    "# take the top three genes in the other category for blast n\n",
    "top_3_in_blastn_other = []\n",
    "increment1 = 0\n",
    "for key,value in sorted(blast_n_other_store.items(),reverse=True):\n",
    "        if increment1 <= 2:\n",
    "                top_3_in_blastn_other.append((key,value))\n",
    "                increment1+=1\n",
    "        else:\n",
    "             break              \n",
    "\n",
    "# take the top three genes in the other category for blast x\n",
    "top_3_in_blastx_other = []\n",
    "increment2 = 0\n",
    "for key,value in sorted(blast_x_other_store.items(),reverse=True):\n",
    "        if increment2 <= 2:\n",
    "                top_3_in_blastx_other.append((key,value))\n",
    "                increment2+=1\n",
    "        else:\n",
    "             break\n",
    "\n",
    "# Store and calculate the ratio of the three selected genes blast n in the other category and plot them onto bar chart\n",
    "blastn_bar_value_count = []\n",
    "blastn_bar_ratios = []\n",
    "blastn_bar_labels = []  \n",
    "blastn_bar_untranslated_labels = []  \n",
    "\n",
    "for idx in range(len(top_3_in_blastn_other)):\n",
    "        blastn_bar_value_count.append(top_3_in_blastn_other[idx][0])\n",
    "        blastn_bar_ratios.append(top_3_in_blastn_other[idx][0]/blast_n_otherCategory_size)\n",
    "        blastn_bar_untranslated_labels.append(top_3_in_blastn_other[idx][1])\n",
    "\n",
    "#replace index sseqids with ncbi name, for blast n bar chart\n",
    "for idx in range(len(blastn_bar_untranslated_labels)):\n",
    "        term = str(blastn_bar_untranslated_labels[idx])\n",
    "        print(term)\n",
    "        esearch_string = esearch(term=term, db='nucleotide')\n",
    "        time.sleep(0.1)\n",
    "        result = get_esummary(esearch_string=esearch_string, db='nucleotide')\n",
    "        result = xmltodict.parse(result)\n",
    "        sseq_name = result['eSummaryResult']['DocumentSummarySet']['DocumentSummary']['Title']\n",
    "        blastn_bar_labels.append(sseq_name)                          \n",
    "\n",
    "        \n",
    "# Store and calculate the ratio of the three selected genes blast x in the other category and plot them onto bar chart\n",
    "blastx_bar_value_count = []   \n",
    "blastx_bar_ratio = []\n",
    "blastx_bar_labels = []\n",
    "blastx_bar_untranslated_labels = []\n",
    "\n",
    "for idx in range(len(top_3_in_blastx_other)):\n",
    "        blastx_bar_ratio.append(top_3_in_blastx_other[idx][0]/blast_x_otherCategory_size)\n",
    "        blastx_bar_untranslated_labels.append(top_3_in_blastx_other[idx][1])\n",
    "        blastx_bar_value_count.append(top_3_in_blastx_other[idx][0])\n",
    "\n",
    "#replace index sseqids with ncbi name, for blast x bar chart\n",
    "for idx in range(len(blastx_bar_untranslated_labels)):\n",
    "        term = str(blastx_bar_untranslated_labels[idx])\n",
    "        print(term)\n",
    "        esearch_string = esearch(term=term, db='nucleotide')\n",
    "        time.sleep(0.3)\n",
    "        result = get_esummary(esearch_string=esearch_string, db='nucleotide')\n",
    "        result = xmltodict.parse(result)\n",
    "        sseq_name = result['eSummaryResult']['DocumentSummarySet']['DocumentSummary']['Title']\n",
    "        blastx_bar_labels.append(sseq_name) \n",
    "                         \n",
    "\n",
    "#remove the low counts genes from the blast n\n",
    "sseq_count_series2 = sseq_count_series1.drop(blast_n_genes_notTop_hit)\n",
    "#remove the low counts genes from the blast x\n",
    "sseq_count_series5 = sseq_count_series4.drop(blast_x_genes_notTop_hit)\n",
    "\n",
    "\n",
    "# Get the gene names for the first pie chart blast n\n",
    "sseq_name_list = []\n",
    "sseqid = []\n",
    "#replace index sseqids with ncbi name\n",
    "for index,values in sseq_count_series2.iteritems():\n",
    "        sseqid.append(index)\n",
    "        term = str(index)\n",
    "        esearch_string = esearch(term=term, db='nucleotide')\n",
    "        time.sleep(0.1)\n",
    "        result = get_esummary(esearch_string=esearch_string, db='nucleotide')\n",
    "        result = xmltodict.parse(result)\n",
    "        sseq_name = result['eSummaryResult']['DocumentSummarySet']['DocumentSummary']['Title']\n",
    "        sseq_name_list.append(sseq_name)\n",
    "        #sseq_count_series2.rename(index={index:sseq_name})\n",
    "#replace the sseqids with ncbi query names\n",
    "replacements = {sseqid:sseq_name_list for sseqid, sseq_name_list in zip(sseqid, sseq_name_list)}\n",
    "sseq_count_series3 = sseq_count_series2.rename(replacements)\n",
    "#sseq_count_series2 = sseq_count_series2.rename(index=dict(zip(sseq_name_list,sseqid)))\n",
    "\n",
    "#generate new pandas series with new element to concatenate with old series\n",
    "d = {'Other':blast_n_otherCategory_size}\n",
    "ser = pd.Series(data=d, index=['Other'])\n",
    "\n",
    "#append new element\n",
    "sseq_count = sseq_count_series3.append(ser)\n",
    "#print(sseq_count)\n",
    "\n",
    "#value_counts returns a pandas series so convert to a data frame\n",
    "sseq_count_df = pd.DataFrame({'sseqid':sseq_count.index, 'count':sseq_count.values})\n",
    "print(sseq_count_df)\n",
    "\n",
    "count = sseq_count_df['count']\n",
    "sseq = sseq_count_df['sseqid']\n",
    "\n",
    "\n",
    "# Get the gene names for the first pie chart blast x\n",
    "sseq_name_list2 = []\n",
    "sseqid2 = []\n",
    "#replace index sseqids with ncbi name\n",
    "for index,values in sseq_count_series5.iteritems():\n",
    "        sseqid2.append(index)\n",
    "        term = str(index)\n",
    "        esearch_string = esearch(term=term, db='nucleotide')\n",
    "        time.sleep(0.1)\n",
    "        result = get_esummary(esearch_string=esearch_string, db='nucleotide')\n",
    "        result = xmltodict.parse(result)\n",
    "        sseq_name = result['eSummaryResult']['DocumentSummarySet']['DocumentSummary']['Title']\n",
    "        sseq_name_list2.append(sseq_name)\n",
    "        #sseq_count_series2.rename(index={index:sseq_name})\n",
    "#replace the sseqids with ncbi query names\n",
    "replacements2 = {sseqid2:sseq_name_list2 for sseqid2, sseq_name_list2 in zip(sseqid2, sseq_name_list2)}\n",
    "sseq_count_series6 = sseq_count_series5.rename(replacements2)\n",
    "#sseq_count_series2 = sseq_count_series2.rename(index=dict(zip(sseq_name_list,sseqid)))\n",
    "\n",
    "#generate new pandas series with new element to concatenate with old series\n",
    "d2 = {'Other':blast_x_otherCategory_size}\n",
    "ser2 = pd.Series(data=d2, index=['Other'])\n",
    "\n",
    "#append new element\n",
    "sseq_count2 = sseq_count_series6.append(ser2)\n",
    "#print(sseq_count)\n",
    "\n",
    "#value_counts returns a pandas series so convert to a data frame\n",
    "sseq_count_df2 = pd.DataFrame({'sseqid':sseq_count2.index, 'count':sseq_count2.values})\n",
    "print(sseq_count_df2)\n",
    "\n",
    "count2 = sseq_count_df2['count']\n",
    "sseq2 = sseq_count_df2['sseqid']\n",
    "\n",
    "\n",
    "\n",
    "# Parsing in the gene labels and counts data to build the plots.\n",
    "print(blastn_bar_value_count)\n",
    "print(blastx_bar_value_count)\n",
    "# Blast n first pie chart and bar chart\n",
    "bottom = 1\n",
    "width = 0.2\n",
    "wedges1, *_ = ax1.pie(count, labels=[f'{sseq[i]}: {count[i]} ({count[i]/sum(count)*100:.1f}% )' for i in range(len(sseq))], colors=None,autopct='',startangle=45,\n",
    "        wedgeprops={\"linewidth\": 1, \"edgecolor\": \"white\"})\n",
    "for j, (height, label, num) in enumerate(reversed([*zip(blastn_bar_ratios, blastn_bar_labels, blastn_bar_value_count)])):\n",
    "    bottom -= height\n",
    "    bc = ax2.bar(0, height, width, bottom=bottom, color='C0', label=label,\n",
    "                 alpha=0.1 + 0.25 * j)\n",
    "    ax2.bar_label(bc, labels=[f\"{height:.0%}({num})\"], label_type='center')\n",
    "ax2.axis('off')\n",
    "ax2.set_xlim(- 2.5 * 2, 2.5 * 2)\n",
    "legend = ax2.legend(loc='upper center', bbox_to_anchor=(0.5, 0.02), ncol=1)\n",
    "\n",
    "\n",
    "# Blast x first pie chart and bar chart\n",
    "bottom = 1\n",
    "width = 0.2\n",
    "wedges2, *_ = ax3.pie(count2, labels=[f'{sseq2[i]}: {count2[i]} ({count2[i]/sum(count2)*100:.1f}% )' for i in range(len(sseq2))], colors=None,autopct='',startangle=45,\n",
    "        wedgeprops={\"linewidth\": 1, \"edgecolor\": \"white\"})\n",
    "for j, (height, label, num) in enumerate(reversed([*zip(blastx_bar_ratio, blastx_bar_labels, blastx_bar_value_count)])):\n",
    "    bottom -= height\n",
    "    bc = ax4.bar(0, height, width, bottom=bottom, color='C0', label=label,\n",
    "                 alpha=0.1 + 0.25 * j)\n",
    "    ax4.bar_label(bc, labels=[f\"{height:.0%}({num})\"], label_type='center')\n",
    "ax4.axis('off')\n",
    "ax4.set_xlim(- 2.5 * 2, 2.5 * 2)\n",
    "legend2 = ax4.legend(loc='upper center', bbox_to_anchor=(0.5, 0.02), ncol=1)\n",
    "\n",
    "\n",
    "### Get the percentage of the the blast n/x mapped results\n",
    "# for ax5\n",
    "ax5_percentages = [10000 - blast_n_mapped_size, blast_n_mapped_size]\n",
    "ax5_labels = ['Unmapped', 'Mapped']\n",
    "ax5.pie(ax5_percentages, labels=[f'{ax5_labels[i]}: {ax5_percentages[i]} ({ax5_percentages[i]/sum(ax5_percentages)*100:.1f}% )' for i in range(len(ax5_labels))], autopct='',\n",
    "       colors=['skyblue', 'gray'], labeldistance=1.1)\n",
    "\n",
    "# for ax6\n",
    "ax6_percentages = [10000 - blast_x_mapped_size, blast_x_mapped_size]\n",
    "ax6_labels = ['Unmapped', 'Mapped']\n",
    "ax6.pie(ax6_percentages, labels=[f'{ax6_labels[i]}: {ax6_percentages[i]} ({ax6_percentages[i]/sum(ax6_percentages)*100:.1f}% )' for i in range(len(ax6_labels))], autopct='',\n",
    "       colors=['skyblue', 'gray'], labeldistance=1.1)\n",
    "\n",
    "# Adding Circle in Pie chart\n",
    "circle1 = plt.Circle((0, 0), radius=0.6, color='white')\n",
    "ax5.add_patch(circle1)\n",
    "circle2 = plt.Circle((0, 0), radius=0.6, color='white')\n",
    "ax6.add_patch(circle2)\n",
    "\n",
    "text1 = str(blast_n_mapped_size) + \" out of 10000 sequences mapped\"\n",
    "text2 = str(blast_x_mapped_size) + \" out of 10000 sequences mapped\"\n",
    "\n",
    "# plt.title('Blastn and Blastx Unmapped Sequences Summary ' + pieName)\n",
    "ax5.text(0.95, 2.2, text1, transform=ax6.transAxes, fontsize=14,\n",
    "        verticalalignment='top', bbox=dict(facecolor='none', edgecolor='black', boxstyle='round,pad=1'))\n",
    "ax6.text(0.95, 0.95, text2, transform=ax6.transAxes, fontsize=14,\n",
    "        verticalalignment='top', bbox=dict(facecolor='none', edgecolor='black', boxstyle='round,pad=1'))\n",
    "\n",
    "fig.tight_layout(pad=5.0)\n",
    "plt.show(block=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "688a5065-699f-498e-a8f9-9211f2a9ffc1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
